docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,543 @@
+
+## Transformation strategy
+
+As mentioned previously, we try to expose as little of this to initial binding as possible. The one major exception to this is our handling of the `RuntimeAsyncMethodAttribute`; we do not let this be applied to
+user code, and will issue an error if a user tries to do this by hand.
+
+Compiler generated async state machines and runtime generated async share some of the same building blocks. Both need to have `await`s with in `catch` and `finally` blocks rewritten to pend the exceptions,
+perform the `await` outside of the `catch`/`finally` region, and then have the exceptions restored as necessary.
+
+TODO: Go over `IAsyncEnumerable` and confirm that the initial rewrite to a `Task`-based method produces code that can then be implemented with runtime async, rather than a full compiler state machine.
+
+TODO: Clarify with the debugger team where NOPs need to be inserted for debugging/ENC scenarios.
+    We will likely need to insert AwaitYieldPoint and AwaitResumePoints for the scenarios where we emit calls to `RuntimeHelpers` async helpers, but can we avoid them for calls in runtime async form?
+
+TODO: Do we need to implement clearing of locals marked with `Hoisted`, or will the runtime handle that?
+
+### Example transformations
+
+Below are some examples of what IL is generated for specific examples.
+
+TODO: Include debug versions
+
+#### General signature transformation
+
+In general, an async method declared in C# will be transformed as follows:
+
+```cs
+async Task M()
+{
+    // ...
+}
+```
+
+```cs
+[RuntimeAsyncMethod, Experimental]
+Task M()
+{
+  // ... see lowering strategy for each kind of await below ...
+}
+```
+
+The same holds for methods that return `Task<T>`, `ValueTask`, and `ValueTask<T>`. Any method returning a different `Task`-like type is not transformed to runtime async form and uses a C#-generated state machine.
+
+`await`s within the body will either be transformed to Runtime-Async call format (as detailed in the runtime specification), or we will use one of the `RuntimeHelpers` methods to do the `await`. Specifics
+for given scenarios are elaborated in more detail below.
+
+`Experimental` will be removed when the full feature is ready to ship, likely not before .NET 11.
+
+TODO: Async iterators (returning `IAsyncEnumerable<T>`)
+
+#### Await `Task`-returning method
+
+```cs
+class C
+{
+    static Task M();
+}
+
+await C.M();
+```
+
+```il
+call modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+```
+
+---------------------------
+
+```cs
+var c = new C();
+await c.M();
+
+class C
+{
+    Task M();
+}
+```
+
+```il
+newobj instance void C::.ctor()
+callvirt instance modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+```
+
+#### Await a concrete `T` `Task<T>`-returning method
+
+```cs
+int i = await C.M();
+
+class C
+{
+    static Task<int> M();
+}
+```
+
+```il
+call modreq(class [System.Runtime]System.Threading.Tasks.Task`1<int32>) int32 C::M()
```

Comment:
This would be just 
```
modreq(class [System.Runtime]System.Threading.Tasks.Task`1)  int32 . . . 
```

I.E. in generic case the modreq uses an open type. 

Otherwise we would need to think about mismatched combinations like `modreq<int> long Foo()` .
Not specifying element type in modreq avoids the problem.

Reply:
If that's the case, then we need to update the runtime document. It currently says this:

> * Otherwise, the return type is the type argument of the return type (either Task`1 or ValueTask`1) with the first custom modifier on the return type being the original type.

I interpret "the original type" to be the full original return type, which is also why I haven't filled in the `Task<T>` generic scenario. If what we really want is the open type, that's perfectly fine, but the spec needs to be clear :slightly_smiling_face:.